File: AdventOfCode2022/Python/Day09/Day09.py
def get_inputs(file):
    """
    function to read in data from a txt file
    """

    with open(file) as f:
        data = f.read().split('\n')

    return data

def Part1(filename):
    """
    Solutions to Part 1
    """
    data = get_inputs(filename)

    H_position = [0,0]
    T_position = [0,0]

    T_positions = [str(T_position[0])+'-'+str(T_position[1])]

    for move in data:

        direction = move.split(' ')[0]
        distance  = int(move.split(' ')[1])

        assert(abs(H_position[0] - T_position[0]) < 2)
        assert(abs(H_position[1] - T_position[1]) < 2)

        for i in range(0,distance):

            if direction == "U":
                H_position[1] += 1

            if direction == "D":
                H_position[1] -= 1

            if direction == "L":
                H_position[0] -= 1

            if direction == "R":
                H_position[0] += 1

            if not ((H_position[0] == T_position[0]) & (H_position[1] == T_position[1])):
     
                # move T up
                if (H_position[0] - T_position[0] == 0) & (H_position[1] - T_position[1] > 1):
                    T_position[1] += 1
                
                # move T down
                if (H_position[0] - T_position[0] == 0) & (H_position[1] - T_position[1] < -1):
                    T_position[1] -= 1

                # move T right
                if (H_position[1] - T_position[1] == 0) & (H_position[0] - T_position[0] >  1):
                    T_position[0] += 1
                
                # move T left
                if (H_position[1] - T_position[1] == 0) & (H_position[0] - T_position[0] < -1):
                    T_position[0] -= 1
                
                # move T up right
                if ((H_position[1] - T_position[1] >= 1) & (H_position[0] - T_position[0] > 1) or
                    (H_position[1] - T_position[1] > 1)  & (H_position[0] - T_position[0] >= 1)):
                    T_position[0] += 1
                    T_position[1] += 1
                
                # move T down right
                if ((H_position[1] - T_position[1] <= -1) & (H_position[0] - T_position[0] > 1) or
                    (H_position[1] - T_position[1] < -1)  & (H_position[0] - T_position[0] >= 1)):
                    T_position[0] += 1
                    T_position[1] -= 1
                
                # move T up left
                if ((H_position[1] - T_position[1] >= 1) & (H_position[0] - T_position[0] < -1) or
                    (H_position[1] - T_position[1] > 1)  & (H_position[0] - T_position[0] <= -1)):
                    T_position[0] -= 1
                    T_position[1] += 1
                
                # move T down left
                if ((H_position[1] - T_position[1] <= -1) & (H_position[0] - T_position[0] < -1) or
                    (H_position[1] - T_position[1] < -1)  & (H_position[0] - T_position[0] <= -1)):
                    T_position[0] -= 1
                    T_position[1] -= 1    

                if str(T_position[0])+'-'+str(T_position[1]) not in T_positions:
                    T_positions.append(str(T_position[0])+'-'+str(T_position[1]))
    
    return len(T_positions)

def Part2(filename):
    """
    Solutions to Part 2
    """
    data = get_inputs(filename)
    
    knot_positions = [[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0],[0,0]]

    T_positions = ['0-0']

    for move in data:

        direction = move.split(' ')[0]
        distance  = int(move.split(' ')[1])

        for i in range(0,distance):

            if direction == "U":
                knot_positions[0][1] += 1

            if direction == "D":
                knot_positions[0][1] -= 1

            if direction == "L":
                knot_positions[0][0] -= 1

            if direction == "R":
                knot_positions[0][0] += 1

            for i in range(1,len(knot_positions)):

                if not ((knot_positions[i-1][0] == knot_positions[i][0]) & (knot_positions[i-1][1] == knot_positions[i][1])):
       
                    # move T up
                    if (knot_positions[i-1][0] - knot_positions[i][0] == 0) & (knot_positions[i-1][1] - knot_positions[i][1] > 1):
                        knot_positions[i][1] += 1
                    
                    # move T down
                    if (knot_positions[i-1][0] - knot_positions[i][0] == 0) & (knot_positions[i-1][1] - knot_positions[i][1] < -1):
                        knot_positions[i][1] -= 1

                    # move T right
                    if (knot_positions[i-1][1] - knot_positions[i][1] == 0) & (knot_positions[i-1][0] - knot_positions[i][0] >  1):
                        knot_positions[i][0] += 1
                    
                    # move T left
                    if (knot_positions[i-1][1] - knot_positions[i][1] == 0) & (knot_positions[i-1][0] - knot_positions[i][0] < -1):
                        knot_positions[i][0] -= 1
                    
                    # move T up right
                    if ((knot_positions[i-1][1] - knot_positions[i][1] >= 1) & (knot_positions[i-1][0] - knot_positions[i][0] > 1) or
                        (knot_positions[i-1][1] - knot_positions[i][1] > 1)  & (knot_positions[i-1][0] - knot_positions[i][0] >= 1)):
                        knot_positions[i][0] += 1
                        knot_positions[i][1] += 1
                    
                    # move T down right
                    if ((knot_positions[i-1][1] - knot_positions[i][1] <= -1) & (knot_positions[i-1][0] - knot_positions[i][0] > 1) or
                        (knot_positions[i-1][1] - knot_positions[i][1] < -1)  & (knot_positions[i-1][0] - knot_positions[i][0] >= 1)):
                        knot_positions[i][0] += 1
                        knot_positions[i][1] -= 1
                    
                    # move T up left
                    if ((knot_positions[i-1][1] - knot_positions[i][1] >= 1) & (knot_positions[i-1][0] - knot_positions[i][0] < -1) or
                        (knot_positions[i-1][1] - knot_positions[i][1] > 1)  & (knot_positions[i-1][0] - knot_positions[i][0] <= -1)):
                        knot_positions[i][0] -= 1
                        knot_positions[i][1] += 1
                    
                    # move T down left
                    if ((knot_positions[i-1][1] - knot_positions[i][1] <= -1) & (knot_positions[i-1][0] - knot_positions[i][0] < -1) or
                        (knot_positions[i-1][1] - knot_positions[i][1] < -1)  & (knot_positions[i-1][0] - knot_positions[i][0] <= -1)):
                        knot_positions[i][0] -= 1
                        knot_positions[i][1] -= 1  

                    if i == len(knot_positions)-1:
                        if str(knot_positions[i][0])+'-'+str(knot_positions[i][1]) not in T_positions:
                            T_positions.append(str(knot_positions[i][0])+'-'+str(knot_positions[i][1]))
                                
            for a in range(1,len(knot_positions)):
                assert(abs(knot_positions[a][0] - knot_positions[a-1][0]) < 2)
                assert(abs(knot_positions[a][1] - knot_positions[a-1][1]) < 2)

    return len(T_positions)

def main():
    """
    main function
    """ 
    testfile = "tests.txt"
    testfile2 = "tests2.txt"
    inputfile = "inputs.txt"

    print("Advent-Of-Code 2022 - Day08")

    print(f'"{testfile}": Answer to Part 1 = {Part1(testfile)}')
    print(f'"{testfile}": Answer to Part 2 = {Part2(testfile)}')

    print(f'"{testfile2}": Answer to Part 1 = {Part1(testfile2)}')
    print(f'"{testfile2}": Answer to Part 2 = {Part2(testfile2)}')

    print(f'"{inputfile}": Answer to Part 1 = {Part1(inputfile)}')
    print(f'"{inputfile}": Answer to Part 2 = {Part2(inputfile)}')

if __name__ == "__main__":
    main()

File: AdventOfCode2022/Go/Day09/Day09.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type coordinate struct {
	x int
	y int
}

type knot struct {
	head coordinate
	tail coordinate
}

func Parse(filename string) []map[string]string {

	file, _ := os.Open(filename)
	fscanner := bufio.NewScanner(file)

	data := []map[string]string{}

	for fscanner.Scan() {
		line := fscanner.Text()
		direction, magnitude := line[:strings.Index(line, " ")], line[strings.Index(line, " ")+1:]

		lineData := map[string]string{"direction": direction, "magnitude": magnitude}
		data = append(data, lineData)
	}

	return data
}

func MoveHead(headPosition coordinate, direction string) coordinate {

	switch direction {
	case "U":
		headPosition.y--
	case "D":
		headPosition.y++
	case "L":
		headPosition.x--
	case "R":
		headPosition.x++
	}

	return headPosition
}

func Catchup(headPosition coordinate, tailPosition coordinate) coordinate {

	newTailPosition := tailPosition

	dx := float64(tailPosition.x - headPosition.x)
	dy := float64(tailPosition.y - headPosition.y)

	theta := math.Atan2(dx, dy)

	if math.Pow(dx, 2)+math.Pow(dy, 2) > 2 {
		newTailPosition.x -= int(math.Round(1.4 * math.Sin(theta)))
		newTailPosition.y -= int(math.Round(1.4 * math.Cos(theta)))
	}

	return newTailPosition
}

func PrintKnots(knots []knot) {

	minX := 0
	maxX := 0

	minY := 0
	maxY := 0

	for _, knot := range knots {
		if knot.head.x < minX {
			minX = knot.head.x
		}
		if knot.head.x > maxX {
			maxX = knot.head.x
		}
		if knot.head.y < minY {
			minY = knot.head.y
		}
		if knot.head.y > maxY {
			maxY = knot.head.y
		}
	}

	canvas := [][]string{}
	for y := minY; y <= maxY+1; y++ {
		row := []string{}
		for x := minX; x <= maxX+1; x++ {
			row = append(row, ".")
		}
		canvas = append(canvas, row)
	}

	for k := 0; k < len(knots); k++ {
		canvas[knots[k].head.y-minY][knots[k].head.x-minX] = strconv.Itoa(k)
	}
	canvas[-minY][-minX] = "s"

	for row := range canvas {
		fmt.Println(canvas[row])
	}
}

func PrintTailPositions(tailPositions map[string]interface{}) {

	minX := 0
	maxX := 0

	minY := 0
	maxY := 0

	for key := range tailPositions {
		x, _ := strconv.Atoi(strings.Split(key, ",")[0])
		y, _ := strconv.Atoi(strings.Split(key, ",")[1])

		if x < minX {
			minX = x
		}
		if x > maxX {
			maxX = x
		}
		if y < minY {
			minY = y
		}
		if y > maxY {
			maxY = y
		}
	}

	canvas := [][]string{}
	for y := minY; y <= maxY+1; y++ {
		row := []string{}
		for x := minX; x <= maxX+1; x++ {
			row = append(row, ".")
		}
		canvas = append(canvas, row)
	}

	for key := range tailPositions {
		x, _ := strconv.Atoi(strings.Split(key, ",")[0])
		y, _ := strconv.Atoi(strings.Split(key, ",")[1])

		canvas[y-minY][x-minX] = "#"
	}

	for row := range canvas {
		fmt.Println(canvas[row])
	}
}

func Part1(filename string) int {

	data := Parse(filename)

	tailPositions := map[string]interface{}{}

	// initialize head position as 0,0
	currentHeadPosition := coordinate{0, 0}
	currentTailPosition := currentHeadPosition

	// draw path
	for _, move := range data {

		direction, magnitudeString := move["direction"], move["magnitude"]
		magnitude, _ := strconv.Atoi(magnitudeString)

		for i := 0; i < magnitude; i++ {

			// move head along
			currentHeadPosition = MoveHead(currentHeadPosition, direction)

			// move tail along
			currentTailPosition = Catchup(currentHeadPosition, currentTailPosition)

			// add tail position to map
			tailPositions[fmt.Sprintf("%v,%v", currentTailPosition.x, currentTailPosition.y)] = ""
		}
	}

	return len(tailPositions)
}

func Part2(filename string) int {

	data := Parse(filename)

	numOfKnots := 9

	tailPositions := map[string]interface{}{}

	// initialize knot positions as 0,0
	knots := []knot{}
	for i := 0; i < numOfKnots; i++ {
		currentHeadPosition := coordinate{0, 0}
		currentTailPosition := currentHeadPosition
		knots = append(knots, knot{head: currentHeadPosition, tail: currentTailPosition})
	}

	// draw path
	for _, move := range data {

		direction, magnitudeString := move["direction"], move["magnitude"]
		// catchupDirection := direction
		magnitude, _ := strconv.Atoi(magnitudeString)

		for i := 0; i < magnitude; i++ {

			// move head along
			knots[0].head = MoveHead(knots[0].head, direction)

			// move tails along
			for k := 0; k < numOfKnots; k++ {
				if k != 0 {
					knots[k].head = knots[k-1].tail
				}
				knots[k].tail = Catchup(knots[k].head, knots[k].tail)
			}

			// add end tail position to map
			tailPositions[fmt.Sprintf("%v,%v", knots[numOfKnots-1].tail.x, knots[numOfKnots-1].tail.y)] = ""

			// PrintKnots(knots)
		}
	}

	// PrintTailPositions(tailPositions)
	return len(tailPositions)
}

func main() {

	testfile := "tests.txt"
	testfile2 := "tests2.txt"
	inputfile := "inputs.txt"

	fmt.Println("Advent-Of-Code 2022 - Day09")
	fmt.Printf("Tests : Answer to Part 1 = %v\n", Part1(testfile))
	fmt.Printf("Inputs: Answer to Part 1 = %v\n", Part1(inputfile))
	fmt.Printf("Tests : Answer to Part 2 = %v\n", Part2(testfile2))
	fmt.Printf("Inputs: Answer to Part 2 = %v\n", Part2(inputfile))

}
